src/tools/info/proto_info.c

```diff
@@ -93,25 +93,211 @@ static void print_resource_usage(const resource_usage_t *usage_before,
+
+    status = ucp_listener_query(listener, &listen_attr);
+    if (status != UCS_OK) {
+        printf("<Failed to query UCP listener>\n");
+        goto out_destroy_listener;
+    }
+
+    status = ucs_sockaddr_get_port((struct sockaddr*)&listen_attr.sockaddr,
+                                   listen_port_p);
+    if (status != UCS_OK) {
+        printf("<Failed to get port>\n");
+        goto out_destroy_listener;
+    }
+
+    *listener_p = listener;
+out:
+    return status;
+
+out_destroy_listener:
+    ucp_listener_destroy(listener);
+    goto out;
+}
+
+ucs_status_t
+print_ucp_ep_info(ucp_worker_h worker, const ucp_ep_params_t *base_ep_params,
+                  const char *ip_addr)
+{
+    ucp_listener_h listener    = NULL;
+    ucp_ep_h server_ep         = NULL;
+    ucp_address_t *worker_addr = NULL;
+    ucp_ep_params_t ep_params  = *base_ep_params;
+    ucs_status_t status, tmp_status;
     ucs_status_ptr_t status_ptr;
+    size_t worker_addr_length;
+    struct sockaddr_in connect_saddr;
+    uint16_t listen_port;
+    ucp_ep_h ep;
+    char ep_name[64];
+    ucp_request_param_t request_param;
+
+    if (ip_addr != NULL) {
+        status = create_listener(worker, &listener, &listen_port, &server_ep);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        ucs_strncpy_zero(ep_name, "client", sizeof(ep_name));
+
+        set_saddr(ip_addr, listen_port, &connect_saddr);
+
+        ep_params.field_mask      |= UCP_EP_PARAM_FIELD_FLAGS |
+                                     UCP_EP_PARAM_FIELD_SOCK_ADDR;
+        ep_params.flags            = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER;
+        ep_params.sockaddr.addr    = (struct sockaddr*)&connect_saddr;
+        ep_params.sockaddr.addrlen = sizeof(connect_saddr);
+    } else {
+        status = ucp_worker_get_address(worker, &worker_addr,
+                                        &worker_addr_length);
+        if (status != UCS_OK) {
+            printf("<Failed to get UCP worker address>\n");
+            return status;
+        }
+
+        ucs_strncpy_zero(ep_name, "connected to UCP worker", sizeof(ep_name));
+
+        ep_params.field_mask |= UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
+        ep_params.address     = worker_addr;
+    }
+
+    status = ucp_ep_create(worker, &ep_params, &ep);
+    if (status != UCS_OK) {
+        printf("<Failed to create UCP endpoint>\n");
+        goto out;
+    }
+
+    request_param.op_attr_mask = 0;
+    /* do EP flush to make sure that fully completed to a peer and final
+     * configuration is applied */
+    status_ptr = ucp_ep_flush_nbx(ep, &request_param);
+    status     = wait_completion(worker, status_ptr);
+    if (status != UCS_OK) {
+        printf("<Failed to flush UCP endpoint>\n");
+    }
+
+    ucp_ep_print_info(ep, stdout);
+
+    tmp_status = ep_close(worker, ep, 0, ep_name);
+    if ((tmp_status != UCS_OK) && (status == UCS_OK)) {
+        status = tmp_status;
+    }
+
+    if (server_ep != NULL) {
```

Comment:
I think we can check ip_addr for consistency with code above and just assert server_ep

Reply:
> (e.g. wrong IP:port specified), so flush completed with error

this case should be handled earlier

> Added ucs_assert(ip_addr != NULL) inside server_ep != NULL

ok